hack/lib.sh

```diff
@@ -0,0 +1,625 @@
+	# push (full $branch_to_push). Assuming the $branch_to_push will be proposed to be merged to
+	# $base_branch_to_diff, so showing a full diff vs $base_branch_to_diff.
+	if upstream_base_head=$(git fetch && git rev-parse "origin/${base_branch_to_diff}"); then
+		if [[ "$(git rev-parse HEAD)" == "${upstream_base_head}" ]]; then
+			echo "⚠️ Nothing to push, even the base origin/${base_branch_to_diff} is up to date; did you expect that?"
+			return 1
+		fi
+		git --no-pager log --oneline "${upstream_base_head}"...HEAD
+		return 0
+	fi
+}
+
+release-lib::exclude_changes_from_last_commit() {
+	local exclude_regexes=$1
+	if [[ -z "${exclude_regexes}" ]]; then
+		echo "❌  exclude_regexes is required." >&2
+		return 1
+	fi
+	local include_regexes=$2
+	if [[ -z "${include_regexes}" ]]; then
+		echo "❌  include_regexes is required." >&2
+		return 1
+	fi
+	local commit_title=$3
+	if [[ -z "${commit_title}" ]]; then
+		echo "❌  commit_title is required." >&2
+		return 1
+	fi
+
+	# Get all files touched by a git commit, delimited by space.
+	changed_files=$(git show --pretty="" --name-only "$(git rev-parse --verify HEAD)")
+	if [ -z "${changed_files}" ]; then
+		echo "❌  suspicious HEAD commit, no files changed." >&2
+		return 1
+	fi
+
+	# Change to \n delimit (needed for grep to work) and exclude/include lines.
+	tmp_to_exclude=$(echo "${changed_files}" | tr ' ' '\n' | grep -E "${exclude_regexes}" | grep -v -E "${include_regexes}")
+
+	# Group node_module and vendor changes, we know we want to get rid of full directories here -- too many of those files slowing things down and obscuring the summary in git commit -- git restore supports globs.
+	to_exclude=$(echo "${tmp_to_exclude}" | gsed -e 's|vendor/.*$|vendor/*|' -e 's|node_modules/.*$|node_modules/*|' | sort -u | tr ' ' '\n')
+	if [ -z "${to_exclude}" ]; then
+		# Nothing to exclude.
+		return 0
+	fi
+
+	echo "🔄 Excluding the following files from the fork squash commit: ${to_exclude}; appending this information to the git commit message"
+	curr_msg=$(git log --format=%B -n1)
+
+	# Get all changes to be in stage area.
+	git reset --soft HEAD~1
+	while IFS= read -r exclude_path; do
+		git restore -S "${exclude_path}"
+	done <<<"${to_exclude}"
+	# Commit after unstaging exclusions.
+	# TODO(bwplotka): Handle nothing to commit after exclusion case.
+	git commit -m "${commit_title}" -m "${curr_msg}" -m "Excluded files:
+${to_exclude}
+"
+	git restore .
+	git clean -fd
+}
+
+release-lib::idemp::dockerfile_update_go_version() {
+	local dockerfile=${1}
+	if [[ -z "${dockerfile}" ]]; then
+		echo "❌  dir arg is required." >&2
+		return 1
+	fi
+
+	if [ ! -f "${dockerfile}" ]; then
+		echo "❌ File not found: $DOCKERFILE"
+		return 1
+	fi
+
+	# TODO test if dockerfile without Go image will fail as expected.
+	local go_version=$(release-lib::dockerfile_go_version "${dockerfile}")
+
+	local golang_tags=$(go tool gcrane ls "google-go.pkg.dev/golang" --json | jq --raw-output '.tags[]' | sort -V)
```

Comment:
This is equivalent to `crane ls google-go.pkg.dev/golang -O | sort -V`, unless you need gcrane for something in particular.

Reply:
Tried to simplify.